server/store/middleware/swarm/swarm.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2016 The Chihaya Authors. All rights reserved.
+// Use of this source code is governed by the BSD 2-Clause license,
+// which can be found in the LICENSE file.
+
+package response
+
+import (
+	"github.com/chihaya/chihaya"
+	"github.com/chihaya/chihaya/pkg/event"
+	"github.com/chihaya/chihaya/server/store"
+	"github.com/chihaya/chihaya/tracker"
+)
+
+func init() {
+	tracker.RegisterAnnounceMiddleware("store_swarm_interaction", announceSwarmInteraction)
+}
+
+// FailedSwarmInteraction represents an error that indicates that the
+// interaction of a peer with a swarm failed.
+type FailedSwarmInteraction string
+
+// Error satisfies the error interface for FailedSwarmInteraction.
+func (f FailedSwarmInteraction) Error() string { return string(f) }
+
+// announceSwarmInteraction provides a middleware that manages swarm
+// interactions for a peer based on the announce.
+func announceSwarmInteraction(next tracker.AnnounceHandler) tracker.AnnounceHandler {
+	return func(cfg *chihaya.TrackerConfig, req *chihaya.AnnounceRequest, resp *chihaya.AnnounceResponse) (err error) {
+		if req.IPv4 != nil {
+			err = updatePeerStore(req, req.Peer4())
+		}
+		if req.IPv6 != nil {
+			err = updatePeerStore(req, req.Peer6())
+		}
+
+		if err != nil {
+			return FailedSwarmInteraction(err.Error())
+		}
+
+		return next(cfg, req, resp)
+	}
+}
+
+func updatePeerStore(req *chihaya.AnnounceRequest, peer chihaya.Peer) (err error) {
+	storage := store.MustGetStore()
+
+	switch req.Event {
+	case event.Completed:
+		err = storage.GraduateLeecher(req.InfoHash, peer)
+	case event.Stopped:
+		if req.Left > 0 {
+			err = storage.DeleteLeecher(req.InfoHash, peer)
+		} else {
+			err = storage.DeleteSeeder(req.InfoHash, peer)
+		}
+	default:
+		if req.Left > 0 {
+			err = storage.PutLeecher(req.InfoHash, peer)
+		} else {
+			err = storage.PutSeeder(req.InfoHash, peer)
+		}
+	}
+
+	return
```

Comment:
here's the code I had mocked up to do similar to this:

```go
func handleEvent(req *chihaya.AnnounceRequest, peer chihaya.Peer) (err error) {
	storage := store.MustGetStore()

	switch {
	case ann.Event == event.Stopped || ann.Event == event.Paused:
		err = storage.DeleteSeeder(req.Infohash, peer)
		if err != nil {
			return err
		}

		err = storage.DeleteLeecher(req.Infohash, peer)
		if err != nil {
			return err
		}

	case ann.Event == event.Completed || req.Left == 0:
		err = storage.GraduateLeecher(req.Infohash, peer)
		if err != nil {
			return err
		}
	}
}
```

In this function I assumed that Delete wouldn't raise an error if you attempted to delete something that was not present. So to be safe, I actually remove the Peer from both seeders and leechers no matter what (since we could've missed the completed event and had them stored as a leecher previously when they're now a seeder). Continuing this same logic, I make completed and all requests with left == 0 use GraduateLeecher since it will guarantee that the peer is no longer in the leechers list if they're a seeder. I haven't looked at this code in a bit, but it looks like if we add a default case that simply sets the peer as a leecher, this should handle all cases.